src/field_theory/splitting_field.lean

```diff
@@ -94,61 +95,405 @@ by rw [nat_degree_remove_factor, hfn, n.add_sub_cancel]
-nat.rec_on n (λ R K _ _ _ _, by exactI ‹algebra R K›) $
-         λ n ih R K _ _ _ f, by exactI ih R
+In order to avoid diamond issues, we have to be careful to define each field of
+an instance of `splitting_field_aux` by recursion, rather than defining the
+whole structure by recursion at once.
+
+We do this in three steps: first define an `add_comm_group` instance to put a `distrib_smul`
+on (which is what is needed to push scalar multiplication through the recursive step),
+then a `comm_ring` instance, and finally a `field` instance.
+-/
+
+/-- Splitting fields have a zero. -/
+protected def zero (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_zero.zero K _) (λ n ih K fK f, ih)
+
+/-- Splitting fields have an addition. -/
+protected def add (n : ℕ) : Π {K : Type u} [field K], by exactI Π {f : K[X]},
+  splitting_field_aux n f → splitting_field_aux n f → splitting_field_aux n f :=
+nat.rec_on n (λ K fK f, by exactI @has_add.add K _) (λ n ih K fK f, ih)
+
+/-- `K` adjoined `n` roots of `f : K[X]` has addition and zero.
+
+We need this instance to define the `distrib_smul` instance below; a `distrib_smul`
+is required to lift scalar multiplications on `K` to a splitting field of `f : K[X]`.
+-/
+instance add_zero_class (n : ℕ) {K : Type u} [field K] {f : K[X]} :
+  add_zero_class (splitting_field_aux n f) :=
+begin
+  refine_struct
+    { add := splitting_field_aux.add n,
+      zero := splitting_field_aux.zero n },
+  all_goals { unfreezingI { induction n with n ih generalizing K } },
+  -- The `succ` cases follow by induction, handle them first.
+  iterate 2 { rotate, exact ih },
+  -- The `zero` cases follow from the structure of the field `K`.
+  all_goals { intros,
+    dsimp only [splitting_field_aux, splitting_field_aux.add, splitting_field_aux.zero],
+    simp },
+end
+
+/-- Splitting fields inherit scalar multiplication. -/
+protected def smul (n : ℕ) : Π (α : Type*) {K : Type u} [field K],
+  by exactI Π [distrib_smul α K],
+  by exactI Π [is_scalar_tower α K K] {f : K[X]},
+  α → splitting_field_aux n f → splitting_field_aux n f :=
+nat.rec_on n
+  (λ α K fK ds ist f, by exactI @has_smul.smul _ K _)
+  (λ n ih α K fK ds ist f, by exactI ih α)
+
+/-- Lift an action of `α` on `K` to `K` adjoined `n` roots of `f : K[X]`.
+
+A `distrib_smul` is required to lift scalar multiplication on `K` to a splitting field of
+`f : K[X]`, i.e. we need this instance to keep the recursive definition going.
+-/
+instance has_smul (α : Type*) (n : ℕ) {K : Type u} [field K] [distrib_smul α K]
+  [is_scalar_tower α K K] {f : K[X]} :
+  has_smul α (splitting_field_aux n f) :=
+{ smul := λ a x, splitting_field_aux.smul n α a x }
+
+/-- Lift an action of `α` on `K` to `K` adjoined `n` roots of `f : K[X]`.
+
+A `distrib_smul` is required to lift scalar multiplication on `K` to a splitting field of
+`f : K[X]`, i.e. we need this instance to keep the recursive definition going.
```

Comment:
Can you point me to where this `distrib_smul` requirement comes from?

Reply:
cc @Vierkantor, I will try and understand this myself but I think she knows better than me exactly what the issue was with doing this without `distrib_smul`. iirc (and it's very faint), it's needed to prove stuff about the Q-action

Reply:
I found the relevant zulip messages:

https://leanprover.zulipchat.com/#narrow/stream/304774-FLT-regular/topic/Project.20status/near/290731108

Essentially, `distrib_mul_action` was not a requirement that could be kept during the induction, and so instead we have this weaker typeclass that can indeed be kept.

Reply:
Is `smul_zero_class` too weak?

Reply:
they were both defined together, so it wasn't an option historically: see #16123. I assume so because a qsmul is always distributive, and so we'd need to keep that.

Reply:
`smul_zero` should be necessary and sufficient to lift scalar multiplication from `K` to `K[X]`. Where we need `smul_add` I believe is in taking the quotient to go from `K[X]` to `adjoin_root (f : K[X])`. Together they give `distrib_smul`.